docs/design/sql/14-automatic-query-resubmission.md

```diff
@@ -0,0 +1,157 @@
+not possible. We could add this option to Java API (to the `SqlStatement`
+class), but this is a non-standard API. We expect many clients to use JDBC or
+other standardized APIs in other languages, which don't provide this support.
+
+2. Another idea is that we could specify this as a query hint, e.g. in the form of
+`SELECT /*+ ALLOW_RETRY */`. But for this we would have to parse the query on
+the client, which we want to avoid. Another reason against this is that queries
+are often generated, such as through ORM or other SQL tools, and injecting the
+hint might not be possible or easy.
+
+3. Another option is to use a session parameter, e.g. executing `SET
+RETRY_STATEGY=<mode>`. The client will have to parse these commands, but we will
+not have to provide a full SQL parser.
+
+4. Provide a new option in `ClientConfig`, and a new JDBC URL parameter for 
+JDBC. Other standardized APIs also provide these. The disadvantage is that it
+will apply to all statements executed using that connection, that is also to DDL
+statements etc.
+
+We propose to implement options (3) and (4). Since the option (4) is feasible
+for all modes, except for not-that-useful `RETRY_ALL` mode, we consider the
+option (3) a nice-to-have feature.
+
+#### Explanation of "no rows received yet"
+
+This means that the `SqlResult.iterator()` didn't ever return from `next()`.
+However, there can be an ongoing blocking call to `next()` in another thread.
+
+#### Other configuration options
+
+The Java client retries invocations in this way: first 5 invocations are retried
+without a delay (hardcoded value). Afterwards the delay is `2^<failureCount>`
+ms, up to 1000 ms (configurable by the
+`hazelcast.client.invocation.retry.pause.millis` property). After 120 seconds
+(configurable by the `hazelcast.client.invocation.timeout.seconds` property), a
+failure is reported to the user.
+
+We think this scheme can be also used for SQL retries and there's no need to
+provide more configuration options.
+
+#### Error classification
+
+Each exception delivered to the client has an [error
+code](https://github.com/hazelcast/hazelcast/blob/713cef1b54b725e6c7df971ff52da30d1133a0a2/hazelcast/src/main/java/com/hazelcast/sql/impl/SqlErrorCode.java#L22).
+This code isn't sufficient to decide whether to retry or not. More error codes
+need to be added. If we retried in case of a problem that has no chance of being
+resolved when executing the query again, it's a waste of time and the error will
+be delivered much later to the user.
+
+Since the use case is to increase availability during member failures, we
+propose to retry only in specific cases (as opposed to _not_ retrying in
+specific cases and retry in all other). We'll rather risk not retrying in a case
+where we could retry, rather than risk retrying in cases where the error could
+be reported immediately.
+
+From the existing codes we need to retry in these cases:
```

Comment:
Are you sure this is all the cases retried by default? We have other cases here at the [shouldRetry](https://github.com/hazelcast/hazelcast/blob/v5.1.2/hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientInvocation.java#L400) method.

I think we should just keep these as it is for all the invocations and we can add more possibilities specific to sql if needed to implement the new retry options.

Reply:
These are the specific cases retried for SQL, not all the cases. Covering all the error cases is a part of the implementation.